test/hooks.js

```diff
@@ -0,0 +1,184 @@
+  const fastify = Fastify()
+
+  t.tearDown(() => fastify.close())
+
+  fastify.register(fastifyWebsocket)
+
+  fastify.addHook('onRequest', async (request, reply) => t.ok('called', 'onRequest'))
+  fastify.addHook('preParsing', async (request, reply, payload) => t.ok('called', 'preParsing'))
+  fastify.addHook('preValidation', async (request, reply) => t.ok('called', 'preValidation'))
+  fastify.addHook('preHandler', async (request, reply) => t.ok('called', 'preHandler'))
+
+  fastify.get('/echo', { websocket: true }, (conn, request) => {
+    conn.setEncoding('utf8')
+    conn.write('hello client')
+    t.tearDown(conn.destroy.bind(conn))
+
+    conn.once('data', chunk => {
+      t.equal(chunk, 'hello server')
+      conn.end()
+    })
+  })
+
+  fastify.listen(0, err => {
+    t.error(err)
+    const ws = new WebSocket('ws://localhost:' + (fastify.server.address()).port + '/echo')
+    const client = WebSocket.createWebSocketStream(ws, { encoding: 'utf8' })
+    t.tearDown(client.destroy.bind(client))
+
+    client.setEncoding('utf8')
+    client.write('hello server')
+
+    client.once('data', chunk => {
+      t.equal(chunk, 'hello client')
+      client.end()
+    })
+  })
+})
+
+test('Should run onTimeout hook', t => {
+  t.plan(3)
+  const fastify = Fastify()
+
+  t.tearDown(() => fastify.close())
+
+  fastify.register(fastifyWebsocket)
+
+  fastify.addHook('onTimeout', async (request, reply) => t.ok('called', 'onTimeout'))
+
+  fastify.get('/echo', { websocket: true }, (conn, request) => {
+    conn.setEncoding('utf8')
+    conn.write('hello client')
+    request.raw.setTimeout(50)
+    t.tearDown(conn.destroy.bind(conn))
+  })
+
+  fastify.listen(0, err => {
+    t.error(err)
+    const ws = new WebSocket('ws://localhost:' + (fastify.server.address()).port + '/echo')
+    const client = WebSocket.createWebSocketStream(ws, { encoding: 'utf8' })
+    t.tearDown(client.destroy.bind(client))
+
+    client.once('data', chunk => {
+      t.equal(chunk, 'hello client')
+    })
+  })
+})
+
+test('Should run onError hook before handler is executed (error thrown in onRequest hook)', t => {
+  t.plan(4)
+  const fastify = Fastify()
+
+  t.tearDown(() => fastify.close())
+
+  fastify.register(fastifyWebsocket)
+
+  const uncaughtException = new Error('uncaught')
+  fastify.addHook('onRequest', async (request, reply) => { throw uncaughtException })
+  fastify.addHook('onError', async (request, reply) => t.ok('called', 'onError'))
+
+  fastify.get('/echo', { websocket: true }, (conn, request) => {
+    t.tearDown(conn.destroy.bind(conn))
+    t.fail('called', 'websocket handler')
+  })
+
+  t.expectUncaughtException(fastify.listen(0, err => {
+    t.error(err)
+    const ws = new WebSocket('ws://localhost:' + (fastify.server.address()).port + '/echo')
+    const client = WebSocket.createWebSocketStream(ws, { encoding: 'utf8' })
+    t.tearDown(client.destroy.bind(client))
+    ws.on('close', code => t.equal(code, 1006))
+  }), uncaughtException)
```

Comment:
Why is `expectUncaughtException` used here?

Reply:
I've used it because when calling `conn.destroy(err)`, this is thrown:
```
events.js:174
      throw er; // Unhandled 'error' event
      ^

Error: uncaught
    at Object.<anonymous> (/Users/ayoub//fastify-websocket/exampl.js:8:27)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
Emitted 'error' event at:
    at Duplex.duplexOnError (/Users/ayoub/fastify-websocket/node_modules/ws/lib/stream.js:37:10)
    at Duplex.emit (events.js:198:13)
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

Using `expectUncaughtException` worked to intercept this for the other node versions that pass the tests.